File: NightConductor/Sources/NightConductor/ConductorDB.swift
import Foundation
import SQLite3

enum ConductorDBError: LocalizedError {
    case notFound(String)
    case sqlite(String)

    var errorDescription: String? {
        switch self {
        case .notFound(let p): return "Conductor database not found at \(p)"
        case .sqlite(let m): return "SQLite: \(m)"
        }
    }
}

/// Read-only scanner for Conductor's session database. A session is
/// "stalled" when its most recent message is a 429 usage-limit error —
/// it died at the limit and nothing has happened since.
enum ConductorDB {
    static var defaultPath: String {
        NSHomeDirectory() + "/Library/Application Support/com.conductor.app/conductor.db"
    }

    private static let lastMessageQuery = """
    WITH last_msgs AS (
        SELECT session_id, content, created_at,
               ROW_NUMBER() OVER (
                   PARTITION BY session_id ORDER BY created_at DESC
               ) AS rn
        FROM session_messages
    )
    SELECT s.id, s.claude_session_id, s.title, w.workspace_path,
           lm.content, lm.created_at
    FROM last_msgs lm
    JOIN sessions s ON s.id = lm.session_id
    JOIN workspaces w ON w.id = s.workspace_id
    WHERE lm.rn = 1
      AND w.state = 'active'
      AND lm.content LIKE '%api_error_status%'
    """

    static func findStalledSessions(dbPath: String = defaultPath) throws -> [StalledSession] {
        guard FileManager.default.fileExists(atPath: dbPath) else {
            throw ConductorDBError.notFound(dbPath)
        }
        var db: OpaquePointer?
        guard sqlite3_open_v2(dbPath, &db, SQLITE_OPEN_READONLY, nil) == SQLITE_OK else {
            throw ConductorDBError.sqlite("cannot open read-only")
        }
        defer { sqlite3_close(db) }

        var statement: OpaquePointer?
        guard sqlite3_prepare_v2(db, lastMessageQuery, -1, &statement, nil) == SQLITE_OK else {
            throw ConductorDBError.sqlite(String(cString: sqlite3_errmsg(db)))
        }
        defer { sqlite3_finalize(statement) }

        var sessions: [StalledSession] = []
        while sqlite3_step(statement) == SQLITE_ROW {
            if let session = parseRow(statement) {
                sessions.append(session)
            }
        }
        return sessions
    }

    private static func parseRow(_ statement: OpaquePointer?) -> StalledSession? {
        func column(_ index: Int32) -> String? {
            guard let text = sqlite3_column_text(statement, index) else { return nil }
            return String(cString: text)
        }
        guard
            let sessionID = column(0),
            let claudeSessionID = column(1), !claudeSessionID.isEmpty,
            let workspacePath = column(3), !workspacePath.isEmpty,
            let content = column(4)
        else { return nil }

        guard
            let data = content.data(using: .utf8),
            let payload = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            payload["type"] as? String == "result",
            payload["is_error"] as? Bool == true,
            (payload["api_error_status"] as? NSNumber)?.intValue == 429
        else { return nil }

        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: workspacePath, isDirectory: &isDirectory),
              isDirectory.boolValue
        else { return nil } // workspace was removed; nothing to resume into

        return StalledSession(
            sessionID: sessionID,
            claudeSessionID: claudeSessionID,
            title: column(2) ?? "Untitled",
            workspacePath: workspacePath,
            errorText: payload["result"] as? String ?? "",
            stalledAt: column(5).flatMap(ISO.parse)
        )
    }
}

File: autoconduct/conductor_db.py
"""Read-only scanner for Conductor's session database.

A session is "stalled" when its most recent *result-type* message is a 429
usage-limit error. (After a 429, Conductor appends synthetic assistant and
system messages, so the error is rarely the literal last row.) The database
is opened with mode=ro; this module never writes.
"""

from __future__ import annotations

import json
import sqlite3
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path

DB_PATH = (
    Path.home()
    / "Library"
    / "Application Support"
    / "com.conductor.app"
    / "conductor.db"
)

# Don't resurrect sessions abandoned at a limit ages ago — they're stale by
# now, and the user has likely moved on.
MAX_STALL_AGE_HOURS = 48

_LAST_RESULT_QUERY = """
WITH results AS (
    SELECT session_id, content, created_at,
           ROW_NUMBER() OVER (
               PARTITION BY session_id ORDER BY created_at DESC, id DESC
           ) AS rn
    FROM session_messages
    WHERE json_valid(content)
      AND json_extract(content, '$.type') = 'result'
)
SELECT s.id, s.claude_session_id, s.title, w.workspace_path,
       r.content, r.created_at
FROM results r
JOIN sessions s ON s.id = r.session_id
JOIN workspaces w ON w.id = s.workspace_id
WHERE r.rn = 1
  AND w.state != 'archived'
  AND s.status != 'working'
  AND json_extract(r.content, '$.is_error') = 1
  AND json_extract(r.content, '$.api_error_status') = 429
"""


@dataclass(frozen=True)
class StalledSession:
    session_id: str
    claude_session_id: str
    title: str
    workspace_path: str
    error_text: str  # e.g. "You've hit your session limit · resets 7:50pm"
    stalled_at: datetime


def _parse_row(row: tuple, now: datetime) -> StalledSession | None:
    session_id, claude_session_id, title, workspace_path, content, created = row
    if not claude_session_id or not workspace_path:
        return None
    if not Path(workspace_path).is_dir():
        return None  # workspace was removed; nothing to resume into

    try:
        payload = json.loads(content)
    except json.JSONDecodeError:
        return None

    stalled_at = datetime.fromisoformat(created.replace("Z", "+00:00"))
    if stalled_at.tzinfo is None:
        stalled_at = stalled_at.replace(tzinfo=timezone.utc)
    if now - stalled_at > timedelta(hours=MAX_STALL_AGE_HOURS):
        return None  # too old; the user has moved on

    return StalledSession(
        session_id=session_id,
        claude_session_id=claude_session_id,
        title=title or "Untitled",
        workspace_path=workspace_path,
        error_text=str(payload.get("result", "")),
        stalled_at=stalled_at,
    )


def find_stalled_sessions(
    db_path: Path = DB_PATH, now: datetime | None = None
) -> tuple[StalledSession, ...]:
    """Return all sessions currently stalled on a usage-limit error."""
    if not db_path.exists():
        raise FileNotFoundError(f"Conductor DB not found at {db_path}")
    reference_time = now or datetime.now(timezone.utc)
    uri = f"file:{db_path}?mode=ro"
    with sqlite3.connect(uri, uri=True, timeout=10) as conn:
        rows = conn.execute(_LAST_RESULT_QUERY).fetchall()
    parsed = (_parse_row(row, reference_time) for row in rows)
    return tuple(s for s in parsed if s is not None)
